src/node.rs

```diff
@@ -1105,8 +1127,21 @@ impl<S: ForkSource + std::fmt::Debug> InMemoryNode<S> {
     }
 
     /// Runs L2 transaction and commits it to a new block.
-    fn run_l2_tx(&self, l2_tx: L2Tx, execution_mode: TxExecutionMode) -> Result<(), String> {
+    fn run_l2_tx(&self, mut l2_tx: L2Tx, execution_mode: TxExecutionMode) -> Result<(), String> {
         let tx_hash = l2_tx.hash();
+        {
+            let inner = self
+                .inner
+                .read()
+                .map_err(|e| format!("Failed to acquire read lock: {}", e))?;
+            if let Some(impersonated_account) = inner.impersonated_account {
```

Comment:
In such a case, you will not be able to use other accounts, when you are impersonating one. This is not compatible with hardhat